lib/clients/statsd.js

```diff
@@ -0,0 +1,79 @@
+var BaseClient = require('./base');
+var util = require('util');
+var types = require('../types');
+var _ = require('lodash');
+var dgram = require('dgram');
+
+var STATS_TYPES = {};
+STATS_TYPES[types.C] = 'c';
+STATS_TYPES[types.G] = 'g';
+STATS_TYPES[types.T] = 'ms';
+
+var KEY_DELIMITER = "-";
+
+function sanatizeInput(input) {
+  //Replace characters that may cause problems here (:| are used by statsd, and cannot be included in the name)
+  return (input + '').replace(/[\s:|=<.]+/g, '');
+}
+
+/**
+ * A client that can send metrics data to a statsd backend
+ * @param {Object} opts options about the statsd backend
+ * @param {String} opts.host the host of the statsd server. Default is 127.0.0.1.
+ * @param {Number} opts.port the port of the statsd server. Default is 8125.
+ */
+function StatsdClient(opts) {
+  BaseClient.apply(this, arguments);
+  opts = opts || {};
+  this.host = opts.host || '127.0.0.1';
+  this.port = opts.port || 8125;
+  this.socket = dgram.createSocket('udp4');
+}
+
+util.inherits(StatsdClient, BaseClient);
+
+//statsd only allows one value per message. If the data contains multiple fields, each of the field will be converted to a message
+//also it doesn't support tags. So we will embed tags into the key. Each message will be constructed like this:
+// <key>-<tagKey>_is_<tagValue>...-<fieldName>:<fieldValue>|<type>
+//For example, given this data:
+// {type: 'gauge', key: 'memory', tags: {host: localhost, workerId:1}, fields: {used: 800, total: 1000}}
+// it will be converted to:
+// ['memory-hostlocalhost-workerId1-used:800|g', 'memory-hostlocalhost-workerId1-total:1000|g']
+StatsdClient.prototype.buildMessages = function(data) {
+  var keys = [];
+  var statsIdentifier = STATS_TYPES[data.type] || 'g';
+  if (data.key) {
+    keys.push(data.key);
+  }
+  //statsd doesn't really support tags, so we just add them as part of the key.
+  var tagsField = data.tags || {};
+  //to make sure the data will show up in the studio, we need to add the `appname` and `statsType` as part of the tags so that it can be used as part of the filter.
+  var appname = process.env.FH_APPNAME || 'NO-APPNAME-DEFINED';
+  var statsType = 'app'; //either `api` or `app`. `api` for stats generated by the fh-mbaas-api
+  tagsField.id = [appname, statsType].join('_');
+  var tags = _.map(tagsField, function(tagValue, tagKey) {
+    return tagKey + '_is_' + tagValue; //see https://github.com/vimeo/carbon-tagger
```

Comment:
![INFO](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-info.png) Move this trailing comment on the previous empty line. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://metrics.skunkhenry.com/coding_rules#rule_key=javascript%3ATrailingComment)